File: src/ruby/request_handler.rb
require(File.expand_path('../json_polyfill.rb', __FILE__))
require(File.expand_path('../action_context.rb', __FILE__))
require(File.expand_path('../utils.rb', __FILE__))

class Bridge

  # @private
  class RequestHandler # interface

    def send(message)
      raise NotImplementedError
    end

    def receive(action_context, request)
      raise NotImplementedError
    end

  end

  # @private
  class DialogRequestHandler < RequestHandler # abstract class

    def initialize(bridge=nil)
      super()
      @bridge = bridge
    end

    def send(message)
      name = message[:name]
      parameters_string = Bridge::JSON.generate(message[:parameters])[1...-1]
      @bridge.dialog.execute_script("#{name}(#{parameters_string})")
    end

    private

    def handle_request(action_context, request)
      unless request.is_a?(Hash) &&
          (defined?(Integer) ? request['id'].is_a?(Integer) : request['id'].is_a?(Fixnum)) &&
          request['name'].is_a?(String) &&
          request['parameters'].is_a?(Array)
        raise(BridgeInternalError, "Bridge received invalid data: \n#{value}")
      end
      id         = request['id']
      name       = request['name']
      parameters = request['parameters'] || []

      # Here we pass a wrapper around the dialog which preserves the message ID to
      # identify the corresponding JavaScript callback.
      # This allows to run asynchronous code (external application etc.) and return
      # later the result to the JavaScript callback even if the dialog has continued
      # sending/receiving messages.
      if request['expectsCallback']
        response = ActionContext.new(@bridge.dialog, self, id)
        # Get the callback.
        unless @bridge.handlers.include?(name)
          raise(BridgeRemoteError.new("No registered callback `#{name}` for #{@bridge.dialog} found."))
        end
        handler = @bridge.handlers[name]
        begin
          handler.call(response, *parameters)
        rescue Exception => error
          # Filter the backtrace if the error was caused in the handler block in another script.
          error.set_backtrace(
            Utils.filter_backtrace(
              error.backtrace,
              exclude_file=__FILE__,
              exclude_line_range=__LINE__-8..__LINE__-2
            )
          )
          # Reject the promise.
          response.reject(error)
          # Re-raise for logging.
          raise(error)
        end
      else
        # Get the callback.
        unless @bridge.handlers.include?(name)
          raise(BridgeRemoteError.new("No registered callback `#{name}` for #{@bridge.dialog} found."))
        end
        handler = @bridge.handlers[name]
        begin
          handler.call(@bridge.dialog, *parameters)
        rescue Exception => error
          # Filter the backtrace if the error was caused in the handler block in another script.
          error.set_backtrace(
            Utils.filter_backtrace(
              error.backtrace,
              exclude_file=__FILE__,
              exclude_line_range=__LINE__-8..__LINE__-2
            )
          )
          if error.is_a?(NoMethodError) && error.message[/undefined method `resolve' for #<UI::(?:Web|Html)Dialog/]
            new_error = NoMethodError.new(
              error.message +
              "\nThe Ruby callback only receives a promise that can be resolved/rejected " +
              "if it is called from JavaScript with Bridge.get('#{name}', …)"
            )
            new_error.set_backtrace(error.backtrace)
            raise(new_error)
          else
            raise(error)
          end
        end
      end
    end

  end

  # @private
  class RequestHandlerHtmlDialog < DialogRequestHandler

    # Receives the raw messages from the HtmlDialog (Bridge.call) and chooses the corresponding callbacks.
    # @private Not for public use.
    # @param   action_context [UI::ActionContext]
    # @param   request        [Object]
    # @private
    def receive(action_context, request)
      handle_request(action_context, request)
    rescue Exception => error
      Utils.log_error(error)
    end

  end

  # @private
  class RequestHandlerWebDialog < DialogRequestHandler

    # Receives the raw messages from the WebDialog (Bridge.call) and chooses the corresponding callbacks.
    # @private Not for public use.
    # @param   dialog           [UI::WebDialog]
    # @param   parameter_string [String]
    def receive(dialog, parameter_string)
      # Get message data from the hidden input element.
      value   = @bridge.dialog.get_element_value("#{NAMESPACE}.requestField") # returns empty string if element not found
      request = Bridge::JSON.parse(value)
      handle_request(dialog, request)
    rescue Exception => error
      Utils.log_error(error)
    ensure
      # Acknowledge that the message has been received and enable the bridge to send
      # the next message if available.
      @bridge.call('Bridge.requestHandler.ack')
    end

  end

end # class Bridge

File: src/js/request-handler.js
export function cleanUpScripts() {
  // Document body may not be loaded yet.
  if (typeof document !== 'undefined' && document.body) {
    let scriptElements = document.body.getElementsByTagName('script')
    for (var i = 0; i < scriptElements.length; i++) {
      scriptElements[i].parentElement.removeChild(scriptElements[i])
    }
  }
}

export class IdGenerator {
  constructor() {
    this.id = 0
  }

  next() {
    // Returns current value, increments variable by one.
    return this.id++
  }
}

function extendError(error, id, handlerName, handler) {
  error.message =
    'Bridge.RequestHandler.receive: Error when executing handler `' +
    handlerName +
    '` (' +
    id +
    '): ' +
    error.message
  if (!error.stack) error.stack = handler.toString()
}

function resolveObjectPath(objectPathString) {
  return objectPathString.split('.').reduce((agg, val) => agg[val], global)
}

function errorToJSON(errorObject) {
  return {
    type: errorObject.name,
    message: errorObject.message,
    backtrace: errorObject.stack,
  }
}

export default class RequestHandler {
  constructor({ messageIdGenerator = new IdGenerator() } = {}) {
    /**
     * A unique identifier for each message. It is used to match return values from Ruby to JavaScript callbacks.
     * @type {number}
     * @private
     * @private
     */
    this.messageIdGenerator = messageIdGenerator

    /**
     * The set of callback handlers waiting to be called from Ruby.
     * @type {object.<number,function>}
     * @private
     */
    this.handlers = {}
  }

  /**
   * The remote end-point calls a JavaScript callback handler.
   * @param  {number} id       The ID of the JavaScript-to-Ruby message.
   * @param  {object} message  An arbitrary object containing the message.
   * @private                  (only for use by corresponding Remote)
   */
  // TODO: Ruby side should pass stack/backtrace or at least caller of Bridge.get or action_context.resolve/reject; source location.
  receive(id, response) {
    // If there is a callback handler, execute it.
    if (this.handlers[id]) {
      let handler = this.handlers[id]
      try {
        if (typeof handler === 'function') {
          handler.call(undefined, response)
        }
      } catch (error) {
        extendError(error, id, handler.name, handler)
        // Log the error.
        //Bridge.error(error) // TODO: remove dependency to Bridge
      }
      delete this.handlers[id]
    } else {
      console.warn(
        `RequestHandler does not know a callback for received response with id=${id}: ${response}`
      )
    }
    cleanUpScripts()
  }

  get(callbackHandlerName, functionName, ...parameters) {
    try {
      let fn = resolveObjectPath(functionName)
      // For easier testing, return the promise (because it is asynchronous).
      // The returned promise is not supposed to be used anywhere else.
      return new Promise(function (resolve, reject) {
        // Call the requested JavaScript function.
        // It may either immediately return a result or a Promise.
        // TODO: If send is asynchronous as well, we need to wrap it into a promise and pass resolve/reject as callback
        resolve(fn.apply(undefined, parameters))
      }).then(
        (result) => {
          return new Promise((resolve, reject) => {
            this.send(
              {
                name: callbackHandlerName,
                parameters: [true, result],
                expectsCallback: false,
              },
              resolve
            )
          })
        },
        (error) => {
          if (error instanceof Error) {
            error = errorToJSON(error)
          }
          this.send({
            name: callbackHandlerName,
            parameters: [false, error],
            expectsCallback: false,
          })
        }
      )
    } catch (error) {
      if (error instanceof Error) {
        error = errorToJSON(error)
      }
      // Reject the promise on the other end-point.
      return new Promise((resolve, reject) => {
        this.send(
          {
            name: callbackHandlerName,
            parameters: [false, error],
            expectsCallback: false,
          },
          resolve
        )
      })
      // Log the error.
      //Bridge.error(error) // TODO: remove dependency to Bridge
    }
  }

  send(message, callback) {
    throw Error('Unimplemented function')
  }
}
